include/exadg/poisson/spatial_discretization/operator.cpp

```diff
@@ -509,26 +526,22 @@ Operator<dim, n_components, Number>::solve(VectorType &       sol,
     check_multigrid.check();
   }
 
-  // Set constrained degrees of freedom of rhs vector according to the prescribed
-  // Dirichlet boundary conditions.
-  VectorType & rhs_mutable = const_cast<VectorType &>(rhs);
+  unsigned int n_iterations = 0;
+
   if(param.spatial_discretization == SpatialDiscretization::CG)
   {
-    laplace_operator.set_constrained_values(rhs_mutable, time);
-  }
-
-  unsigned int iterations = iterative_solver->solve(sol, rhs_mutable);
+    n_iterations = iterative_solver->solve(sol, rhs);
 
-  // This step should actually be optional: The constrained degrees of freedom of the
-  // rhs vector contain the Dirichlet boundary values and the linear operator contains
-  // values of 1 on the diagonal. Hence, sol should already contain the correct
-  // Dirichlet boundary values for constrained degrees of freedom.
-  if(param.spatial_discretization == SpatialDiscretization::CG)
+    // Set Dirichlet degrees of freedom according to Dirichlet boundary condition.
+    laplace_operator.set_time(time);
+    laplace_operator.set_inhomogeneous_boundary_values(sol);
+  }
+  else
   {
-    laplace_operator.set_constrained_values(sol, time);
+    n_iterations = iterative_solver->solve(sol, rhs);
   }
 
```

Comment:
This definitely closes the issue #446 :+1: . Setting the boundary values after the solution step (in contrast to before the solution step in the previous implementation) is intended right?

Apart from that I suggest writing:

```cpp
unsigned int const n_iterations = iterative_solver->solve(sol, rhs);

if (param.spatial_discretization == SpatialDiscretization::CG)
{
    // Set Dirichlet degrees of freedom according to Dirichlet boundary condition.
    laplace_operator.set_time(time);
    laplace_operator.set_inhomogeneous_boundary_values(sol);
}

return n_iterations;
```

Reply:
Yes, this is intended. Before, we have actually set the dofs twice. But there should actually be no need to do it before solving (because the inhomogeneous parts have been shifted to the rhs and we only solve for the non-constrained dofs; while constrained dofs are set after solving). So I think we should avoid such non-unique logic that we had before.

Thanks for the suggestion! Several changes ended up in an ugly code ;)